internal/messagequeue/donthavetimeoutmgr.go

```diff
@@ -0,0 +1,297 @@
+
+	// Check if after changing the timeout there are any pending wants that are
+	// now over the timeout
+	dhtm.checkForTimeouts()
+}
+
+// checkForTimeouts checks pending wants to see if any are over the timeout
+func (dhtm *dontHaveTimeoutMgr) checkForTimeouts() {
+	dhtm.lk.Lock()
+	defer dhtm.lk.Unlock()
+
+	if len(dhtm.wantQueue) == 0 {
+		return
+	}
+
+	// Figure out which of the blocks that were wanted were not received
+	// within the timeout
+	more := true
+	expired := make([]cid.Cid, 0, len(dhtm.pending))
+	removeCount := 0
+	for i := 0; i < len(dhtm.wantQueue) && more; i++ {
+		c := dhtm.wantQueue[i]
+
+		// If the want has been cancelled, its cid will have been removed from
+		// the pending map
+		if started, ok := dhtm.pending[c]; ok {
+			// If the want has expired
+			if time.Since(started) > dhtm.timeout {
+				// Add the want to the expired list
+				expired = append(expired, c)
+				// Remove the want from the pending wants
+				delete(dhtm.pending, c)
+				// Remove expired wants from the want queue
+				removeCount++
+			} else {
+				// The queue is in order from earliest to latest, so if we
+				// didn't find an expired entry we can stop iterating
+				more = false
+			}
+		} else {
+			// Remove cancelled wants from the want queue
+			removeCount++
+		}
+	}
+
+	// Remove cancelled or expired wants from want queue
+	dhtm.wantQueue = dhtm.wantQueue[removeCount:]
+
+	// Fire the timeout event for the expired wants
+	if len(expired) > 0 {
+		go dhtm.fireTimeout(expired)
+	}
+
+	if len(dhtm.wantQueue) == 0 {
+		return
+	}
+
+	// Make sure the timeout manager is still running
+	if dhtm.ctx.Err() != nil {
+		return
+	}
+
+	// Note: we cleared cancelled wants from the front of the queue above
+	oldest, ok := dhtm.pending[dhtm.wantQueue[0]]
+	if !ok {
+		panic("cancelled want found in DONT_HAVE timeout manager want queue after purge")
+	}
+
+	// Schedule the next check for the moment when the oldest pending want will
+	// timeout
+	until := time.Until(oldest.Add(dhtm.timeout))
+	if dhtm.checkForTimeoutsTimer == nil {
+		dhtm.checkForTimeoutsTimer = time.AfterFunc(until, dhtm.checkForTimeouts)
+	} else {
+		dhtm.checkForTimeoutsTimer.Stop()
+		dhtm.checkForTimeoutsTimer.Reset(until)
+	}
+}
+
+// AddPending adds the given keys that will expire if not cancelled before
+// the timeout
+func (dhtm *dontHaveTimeoutMgr) AddPending(ks []cid.Cid) {
+	if len(ks) == 0 {
+		return
+	}
+
+	start := time.Now()
+
+	dhtm.lk.Lock()
+
+	queueWasEmpty := len(dhtm.pending) == 0
+
+	// Record the start time for each key
+	for _, c := range ks {
+		if _, ok := dhtm.pending[c]; !ok {
+			dhtm.pending[c] = start
+			dhtm.wantQueue = append(dhtm.wantQueue, c)
+		}
+	}
+
+	dhtm.lk.Unlock()
+
+	// If there was already an earlier pending item in the queue, then there
+	// must already be a timeout check scheduled. If there is nothing in the
+	// queue then we should make sure to schedule a check.
+	if queueWasEmpty {
+		dhtm.checkForTimeouts()
```

Comment:
nit: this is a little funky given that 2/3 times we use this function, we have the lock, drop it, then re-take it.